NPE occurs during trace of ManagedExecutorServiceImpl

In slack thread https://ibm-cloud.slack.com/archives/C30NGTBFS/p1622042314179900, an IBM cloud customer posted a stack trace from an issue he was having with some Db2 config and the Db2 cloud service.  The stack trace showed an NPE in ManagedExecutorServiceImpl. toString(), which appears to need null checks for either the name and hashcode fields:

```
CWWKE0701E: bundle com.ibm.ws.concurrent:1.0.47.cl201220201111-0736 (184)[com.ibm.ws.concurrent.internal.ManagedExecutorServiceImpl(416)] : Error during instantiation of the implementation object java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:83)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:57)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:437)
	at org.apache.felix.scr.impl.inject.internal.ComponentConstructorImpl.newInstance(ComponentConstructorImpl.java:312)
	at org.apache.felix.scr.impl.manager.SingleComponentManager.createImplementationObject(SingleComponentManager.java:286)
	at org.apache.felix.scr.impl.manager.SingleComponentManager.createComponent(SingleComponentManager.java:115)
	at org.apache.felix.scr.impl.manager.SingleComponentManager.getService(SingleComponentManager.java:1000)
	at org.apache.felix.scr.impl.manager.SingleComponentManager.getServiceInternal(SingleComponentManager.java:973)
	at org.apache.felix.scr.impl.manager.SingleComponentManager.getService(SingleComponentManager.java:918)
	at org.eclipse.osgi.internal.serviceregistry.ServiceFactoryUse$1.run(ServiceFactoryUse.java:216)
	at java.security.AccessController.doPrivileged(AccessController.java:678)
	at org.eclipse.osgi.internal.serviceregistry.ServiceFactoryUse.factoryGetService(ServiceFactoryUse.java:213)
	at org.eclipse.osgi.internal.serviceregistry.ServiceFactoryUse.getService(ServiceFactoryUse.java:114)
	at org.eclipse.osgi.internal.serviceregistry.ServiceConsumer$2.getService(ServiceConsumer.java:48)
	at org.eclipse.osgi.internal.serviceregistry.ServiceRegistrationImpl.getService(ServiceRegistrationImpl.java:530)
	at org.eclipse.osgi.internal.serviceregistry.ServiceRegistry.getService(ServiceRegistry.java:494)
	at org.eclipse.osgi.internal.framework.BundleContextImpl.getService(BundleContextImpl.java:685)
	at org.apache.felix.scr.impl.manager.SingleRefPair.getServiceObject(SingleRefPair.java:88)
	at org.apache.felix.scr.impl.inject.methods.BindMethod.getServiceObject(BindMethod.java:675)
	at org.apache.felix.scr.impl.manager.DependencyManager.getServiceObject(DependencyManager.java:2545)
	at org.apache.felix.scr.impl.manager.DependencyManager.doInvokeBindMethod(DependencyManager.java:2064)
	at org.apache.felix.scr.impl.manager.DependencyManager.invokeBindMethod(DependencyManager.java:2047)
	at org.apache.felix.scr.impl.manager.SingleComponentManager.invokeBindMethod(SingleComponentManager.java:443)
	at org.apache.felix.scr.impl.manager.DependencyManager$MultipleDynamicCustomizer.addedService(DependencyManager.java:333)
	at org.apache.felix.scr.impl.manager.DependencyManager$MultipleDynamicCustomizer.addedService(DependencyManager.java:301)
	at org.apache.felix.scr.impl.manager.ServiceTracker$Tracked.customizerAdded(ServiceTracker.java:1200)
	at org.apache.felix.scr.impl.manager.ServiceTracker$Tracked.customizerAdded(ServiceTracker.java:1121)
	at org.apache.felix.scr.impl.manager.ServiceTracker$AbstractTracked.trackAdding(ServiceTracker.java:928)
	at org.apache.felix.scr.impl.manager.ServiceTracker$AbstractTracked.track(ServiceTracker.java:864)
	at org.apache.felix.scr.impl.manager.ServiceTracker$Tracked.serviceChanged(ServiceTracker.java:1152)
	at org.apache.felix.scr.impl.BundleComponentActivator$ListenerInfo.serviceChanged(BundleComponentActivator.java:114)
	at org.eclipse.osgi.internal.serviceregistry.FilteredServiceListener.serviceChanged(FilteredServiceListener.java:113)
	at org.eclipse.osgi.internal.framework.BundleContextImpl.dispatchEvent(BundleContextImpl.java:985)
	at org.eclipse.osgi.framework.eventmgr.EventManager.dispatchEvent(EventManager.java:234)
	at org.eclipse.osgi.framework.eventmgr.ListenerQueue.dispatchEventSynchronous(ListenerQueue.java:151)
	at org.eclipse.osgi.internal.serviceregistry.ServiceRegistry.publishServiceEventPrivileged(ServiceRegistry.java:896)
	at org.eclipse.osgi.internal.serviceregistry.ServiceRegistry.publishServiceEvent(ServiceRegistry.java:834)
	at org.eclipse.osgi.internal.serviceregistry.ServiceRegistrationImpl.register(ServiceRegistrationImpl.java:130)
	at org.eclipse.osgi.internal.serviceregistry.ServiceRegistry.registerService(ServiceRegistry.java:258)
	at org.eclipse.osgi.internal.framework.BundleContextImpl.registerService(BundleContextImpl.java:525)
	at org.apache.felix.scr.impl.manager.AbstractComponentManager$3.register(AbstractComponentManager.java:929)
	at org.apache.felix.scr.impl.manager.AbstractComponentManager$3.register(AbstractComponentManager.java:915)
	at org.apache.felix.scr.impl.manager.RegistrationManager.changeRegistration(RegistrationManager.java:133)
	at org.apache.felix.scr.impl.manager.AbstractComponentManager.registerService(AbstractComponentManager.java:984)
	at org.apache.felix.scr.impl.manager.AbstractComponentManager.activateInternal(AbstractComponentManager.java:752)
	at org.apache.felix.scr.impl.manager.AbstractComponentManager.enableInternal(AbstractComponentManager.java:674)
	at org.apache.felix.scr.impl.manager.AbstractComponentManager.enable(AbstractComponentManager.java:437)
	at org.apache.felix.scr.impl.manager.ConfigurableComponentHolder.enableComponents(ConfigurableComponentHolder.java:667)
	at org.apache.felix.scr.impl.BundleComponentActivator.initialEnable(BundleComponentActivator.java:305)
	at org.apache.felix.scr.impl.Activator.loadComponents(Activator.java:554)
	at org.apache.felix.scr.impl.Activator.access$200(Activator.java:70)
	at org.apache.felix.scr.impl.Activator$ScrExtension.start(Activator.java:421)
	at org.apache.felix.scr.impl.AbstractExtender.createExtension(AbstractExtender.java:196)
	at org.apache.felix.scr.impl.AbstractExtender.modifiedBundle(AbstractExtender.java:169)
	at org.apache.felix.scr.impl.AbstractExtender.modifiedBundle(AbstractExtender.java:49)
	at org.osgi.util.tracker.BundleTracker$Tracked.customizerModified(BundleTracker.java:488)
	at org.osgi.util.tracker.BundleTracker$Tracked.customizerModified(BundleTracker.java:1)
	at org.osgi.util.tracker.AbstractTracked.track(AbstractTracked.java:232)
	at org.osgi.util.tracker.BundleTracker$Tracked.bundleChanged(BundleTracker.java:450)
	at org.eclipse.osgi.internal.framework.BundleContextImpl.dispatchEvent(BundleContextImpl.java:973)
	at org.eclipse.osgi.framework.eventmgr.EventManager.dispatchEvent(EventManager.java:234)
	at org.eclipse.osgi.framework.eventmgr.ListenerQueue.dispatchEventSynchronous(ListenerQueue.java:151)
	at org.eclipse.osgi.internal.framework.EquinoxEventPublisher.publishBundleEventPrivileged(EquinoxEventPublisher.java:235)
	at org.eclipse.osgi.internal.framework.EquinoxEventPublisher.publishBundleEvent(EquinoxEventPublisher.java:141)
	at org.eclipse.osgi.internal.framework.EquinoxEventPublisher.publishBundleEvent(EquinoxEventPublisher.java:133)
	at org.eclipse.osgi.internal.framework.EquinoxContainerAdaptor.publishModuleEvent(EquinoxContainerAdaptor.java:217)
	at org.eclipse.osgi.container.Module.publishEvent(Module.java:498)
	at org.eclipse.osgi.container.Module.start(Module.java:485)
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel$2.run(ModuleContainer.java:1844)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1160)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.lang.Thread.run(Thread.java:820)
Caused by: java.lang.NullPointerException
	at com.ibm.ws.concurrent.internal.ManagedExecutorServiceImpl.toString(ManagedExecutorServiceImpl.java:774)
	at com.ibm.ws.logging.hpel.impl.BinaryLogRecordSerializerVersion2Impl.writeParam(BinaryLogRecordSerializerVersion2Impl.java:781)
	at com.ibm.ws.logging.hpel.impl.BinaryLogRecordSerializerVersion2Impl.serialize(BinaryLogRecordSerializerVersion2Impl.java:311)
	at com.ibm.ws.logging.hpel.handlers.LogRecordHandler$SerializationBuffer.serialize(LogRecordHandler.java:190)
	at com.ibm.ws.logging.hpel.handlers.LogRecordHandler.processEvent(LogRecordHandler.java:95)
	at com.ibm.ws.logging.hpel.handlers.LogRecordHandler.publish(LogRecordHandler.java:236)
	at com.ibm.ws.logging.internal.hpel.HpelTraceServiceWriter$NoCloseLogRecordHandler.publish(HpelTraceServiceWriter.java:305)
	at com.ibm.ws.logging.internal.hpel.HpelTraceServiceWriter.repositoryPublish(HpelTraceServiceWriter.java:283)
	at com.ibm.ws.logging.internal.hpel.HpelBaseTraceService.publishTraceLogRecord(HpelBaseTraceService.java:131)
	at com.ibm.ws.logging.internal.impl.BaseTraceService.entry(BaseTraceService.java:745)
	at com.ibm.websphere.ras.Tr.entry(Tr.java:483)
	at com.ibm.ws.concurrent.internal.ManagedExecutorServiceImpl$2.<init>(ManagedExecutorServiceImpl.java:132)
	at com.ibm.ws.concurrent.internal.ManagedExecutorServiceImpl.<init>(ManagedExecutorServiceImpl.java:132)
	... 73 more
  ,formattedMsg=CWWKE0701E: bundle com.ibm.ws.concurrent:1.0.47.cl201220201111-0736 (184)[com.ibm.ws.concurrent.internal.ManagedExecutorServiceImpl(416)] : Error during instantiation of the implementation object java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:83)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:57)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:437)
	at org.apache.felix.scr.impl.inject.internal.ComponentConstructorImpl.newInstance(ComponentConstructorImpl.java:312)
	at [internal classes]
Caused by: java.lang.NullPointerException
	at com.ibm.ws.concurrent.internal.ManagedExecutorServiceImpl.toString(ManagedExecutorServiceImpl.java:774)
	at [internal classes]
	at com.ibm.ws.logging.hpel.impl.BinaryLogRecordSerializerVersion2Impl.writeParam(BinaryLogRecordSerializerVersion2Impl.java:781)
	at com.ibm.ws.logging.hpel.impl.BinaryLogRecordSerializerVersion2Impl.serialize(BinaryLogRecordSerializerVersion2Impl.java:311)
	at com.ibm.ws.logging.hpel.handlers.LogRecordHandler$SerializationBuffer.serialize(LogRecordHandler.java:190)
	at com.ibm.ws.logging.hpel.handlers.LogRecordHandler.processEvent(LogRecordHandler.java:95)
	at com.ibm.ws.logging.hpel.handlers.LogRecordHandler.publish(LogRecordHandler.java:236)
	at com.ibm.ws.logging.internal.hpel.HpelTraceServiceWriter$NoCloseLogRecordHandler.publish(HpelTraceServiceWriter.java:305)
	at com.ibm.ws.logging.internal.hpel.HpelTraceServiceWriter.repositoryPublish(HpelTraceServiceWriter.java:283)
	at com.ibm.ws.logging.internal.hpel.HpelBaseTraceService.publishTraceLogRecord(HpelBaseTraceService.java:131)
	at com.ibm.ws.logging.internal.impl.BaseTraceService.entry(BaseTraceService.java:745)
	at com.ibm.websphere.ras.Tr.entry(Tr.java:483)
	... 1 more
```

--- a/dev/com.ibm.ws.concurrent/src/com/ibm/ws/concurrent/internal/ManagedExecutorServiceImpl.java
+++ b/dev/com.ibm.ws.concurrent/src/com/ibm/ws/concurrent/internal/ManagedExecutorServiceImpl.java
@@ -127,21 +127,6 @@ public class ManagedExecutorServiceImpl implements ExecutorService, //
     private ConcurrencyService concurrencySvc;
 
     /**
-     * Privileged action to lazily obtain the context service. Available only on the OSGi code path.
-     */
-    private final PrivilegedAction<WSContextService> contextSvcAccessor = new PrivilegedAction<WSContextService>() {
-        @Override
-        @Trivial
-        public WSContextService run() {
-            try {
-                return contextSvcRef.getServiceWithException();
-            } catch (IllegalStateException x) {
-                throw new RejectedExecutionException(x);
-            }
-        }
-    };
-
-    /**
      * Reference to the context service for this managed executor service. Available only on the OSGi code path.
      */
     private final AtomicServiceReference<WSContextService> contextSvcRef = new AtomicServiceReference<WSContextService>("ContextService");
@@ -192,17 +177,6 @@ public class ManagedExecutorServiceImpl implements ExecutorService, //
      * Executor that runs tasks against the general concurrency policy for this managed executor.
      */
     volatile PolicyExecutor policyExecutor;
-
-    /**
-     * Privileged action to lazily obtain the transaction context provider.
-     */
-    private final PrivilegedAction<ThreadContextProvider> tranContextProviderAccessor = new PrivilegedAction<ThreadContextProvider>() {
-        @Override
-        @Trivial
-        public ThreadContextProvider run() {
-            return tranContextProviderRef.getService();
-        }
-    };
 
     /**
      * Reference to the transaction context provider.
@@ -444,7 +418,14 @@ public class ManagedExecutorServiceImpl implements ExecutorService, //
 
     @Override
     public WSContextService getContextService() {
-        return mpContextService == null ? AccessController.doPrivileged(contextSvcAccessor) : mpContextService;
+        if (mpContextService == null)
+            try {
+                return contextSvcRef.getServiceWithException(); // doPriv is covered by AtomicServiceReference
+            } catch (IllegalStateException x) {
+                throw new RejectedExecutionException(x);
+            }
+        else
+            return mpContextService;
     }
 
     @Override
@@ -766,7 +747,7 @@ public class ManagedExecutorServiceImpl implements ExecutorService, //
     @SuppressWarnings("deprecation")
     ThreadContext suspendTransaction() {
         Map<String, String> XPROPS_SUSPEND_TRAN = eeVersion < 9 ? JAVAX_SUSPEND_TRAN : JAKARTA_SUSPEND_TRAN;
-        ThreadContextProvider tranContextProvider = AccessController.doPrivileged(tranContextProviderAccessor);
+        ThreadContextProvider tranContextProvider = tranContextProviderRef.getService(); // doPriv is covered by AtomicServiceReference
         ThreadContext suspendedTranSnapshot = tranContextProvider == null ? null : tranContextProvider.captureThreadContext(XPROPS_SUSPEND_TRAN, null);
         if (suspendedTranSnapshot != null)
             suspendedTranSnapshot.taskStarting();
@@ -777,7 +758,7 @@ public class ManagedExecutorServiceImpl implements ExecutorService, //
     @Trivial
     public String toString() {
         String s = name.get();
-        return s.startsWith("ManagedExecutor@") ? s : ("ManagedExecutor@" + Integer.toHexString(hashCode()) + ' ' + s);
+        return s != null && s.startsWith("ManagedExecutor@") ? s : ("ManagedExecutor@" + Integer.toHexString(hashCode()) + ' ' + s);
     }
 
     /**

